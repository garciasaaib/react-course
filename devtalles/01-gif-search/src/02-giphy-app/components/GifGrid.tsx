import { GifItem } from "./";
import { useFetchGifs } from "../hooks/useFetchGifs";

export const GifGrid = ({ category }: { category: string }) => {
	const { gifs, loading } = useFetchGifs(category);

	return (
		<>
			<h3>{category}</h3>
			{loading && <h2>Cargando...</h2>}

			<div className="card-grid">
				{gifs.map((image) => (
					<GifItem key={image.id} {...image} />
				))}
			</div>
		</>
	);
};
