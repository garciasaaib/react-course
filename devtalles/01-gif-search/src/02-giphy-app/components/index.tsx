export { AddCategory } from './AddCategory';
export { GifGrid } from './GifGrid';
export { GifItem } from './GifItem';