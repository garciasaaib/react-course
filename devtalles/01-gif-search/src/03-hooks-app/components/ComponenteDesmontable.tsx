import { useEffect, useState } from "react"

export const ComponenteDesmontable = () => {
  const [coords, setCoords] = useState({x:0, y:0})
  // reglas de montaje y desmontaje
  useEffect(() => {
    console.log('Message Mounted')
    return () => {
      console.log('Message Unmounted')
    }
  }, [])

  // efecto encadenado por el movimiento del mouse
  useEffect(() => {
    // creamos la funcion
    const onMouseMove = ({x, y}: MouseEvent) => {
      const coords = {x,y}
      setCoords(coords)
    }

    // agregamos esta funcion por referencia cuando se monta
    window.addEventListener('mousemove', onMouseMove)

    // quitamos la funcion del evento por referencia
    return () => {
      window.removeEventListener('mousemove',onMouseMove)
    }
  },[])
  return (
    <>
      <h3>Usuario ya existe</h3>
      <p>x: {coords.x}</p>
      <p>y: {coords.y}</p>
    </>
  )
}
