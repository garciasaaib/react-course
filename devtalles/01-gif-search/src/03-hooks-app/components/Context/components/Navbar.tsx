import { Link, useLocation } from "react-router";

export const Navbar = () => {
  const {pathname} = useLocation()
  const onIsActive = (thisPath: string) => {
    return pathname === thisPath ? " underline" : "";

  }
	return (
		<nav className="flex gap-1">
			<Link className={`hover:text-white${onIsActive('/')}`} to="/">
				Home
			</Link>
			<Link className={`hover:text-white${onIsActive('/about')}`} to="/about">
				About
			</Link>
			<Link className={`hover:text-white${onIsActive('/login')}`} to="/login">
				Login
			</Link>
		</nav>
	);
};
