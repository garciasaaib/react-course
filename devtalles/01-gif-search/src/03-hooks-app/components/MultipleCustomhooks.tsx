export const MultipleCustomhooks = () => {
  return (
    <div>MultipleCustomhooks</div>
  )
}