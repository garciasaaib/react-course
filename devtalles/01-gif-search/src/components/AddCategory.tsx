import { useState } from 'react';

export const AddCategory = ({ onNewCategory }: { onNewCategory: (newCategory:string) => void}) => {

    const [ inputValue, setInputValue ] = useState('');

    const onInputChange = ({ target }: React.ChangeEvent<HTMLInputElement>) => {
        setInputValue( target.value );
    }

    const onSubmit = ( event: React.FormEvent<HTMLFormElement> ) => {
        event.preventDefault();
        if( inputValue.trim().length <= 1) return;

        // setCategories( categories => [ inputValue, ...categories ]);
        setInputValue('');
        onNewCategory( inputValue.trim() );
    }

    return (
        <form onSubmit={onSubmit} aria-label="form">
            <input
                type="text"
                placeholder="Buscar gifs"
                value={ inputValue }
                onChange={onInputChange}
            />
        </form>
    )
}
