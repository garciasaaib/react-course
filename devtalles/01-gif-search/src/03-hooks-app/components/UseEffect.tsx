import { ChangeEvent, useEffect, useState } from "react";
import { ComponenteDesmontable } from "./ComponenteDesmontable";

export const UseEffect = () => {
  const [state, setState] = useState({
    username: 'Adrian',
    email: 'adrian@gmail.com',
  })
  const {username, email} = state
  const onInputChange = ({target}: ChangeEvent<HTMLInputElement>) => {
    const {value, name} = target
    setState({...state, [name]: value})
  }
  useEffect(() => {
    // body: all that this does


    // return () => {
      // clean up: cleanning for unmount component
    //}
  }, []) // dependencies: watching variables
	return (
		<>
			<h1>SimpleForm</h1>
			<hr />
			<form>
				<input onChange={onInputChange} value={username} type="text" placeholder="Username" name="username" />
				<input onChange={onInputChange} value={email} type="email" placeholder="Email" name="email" />
			</form>
      {/* El monte y desmonte se genera desde el compoonente padre */}
      { username === 'Adrian' && <ComponenteDesmontable /> }
		</>
	);
};
