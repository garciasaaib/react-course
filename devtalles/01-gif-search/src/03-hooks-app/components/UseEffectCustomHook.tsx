import { ComponenteDesmontable } from "./ComponenteDesmontable";
import { useEffectCustomHook } from "../hooks/useEffectCustomHook";

const INITIAL_VALUES = {
	username: "",
	email: "",
	password: "",
};
export const UseEffectCustomHook = () => {
	const { state, onInputChange, onInputReset } = useEffectCustomHook({ ...INITIAL_VALUES });

	return (
		<>
			<h1>SimpleForm</h1>
			<hr />
			<form>
				<input
					onChange={onInputChange}
					value={state.username}
					type="text"
					placeholder="Username"
					name="username"
				/>
				<input
					onChange={onInputChange}
					value={state.email}
					type="email"
					placeholder="Email"
					name="email"
				/>
				<input
					onChange={onInputChange}
					value={state.password}
					type="password"
					placeholder="Password"
					name="password"
				/>
			</form>
      <button className="btn" onClick={onInputReset}>Reset</button>
			{/* El monte y desmonte se genera desde el compoonente padre */}
			{state.username === "Adrian" && <ComponenteDesmontable />}
		</>
	);
};
