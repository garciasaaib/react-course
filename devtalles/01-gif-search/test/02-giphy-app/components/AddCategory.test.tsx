import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import { AddCategory } from "../../../src/02-giphy-app/components/AddCategory";
const newCategory = (data: string) => {
	return data;
};
describe("<AddCategory />", () => {
	test("Debe crear en snapshoot de AddCategory", () => {
		const { container } = render(<AddCategory onNewCategory={newCategory} />);
		// solo crea el snapshot
		expect(container).toMatchSnapshot();
	});

	test("Debe crear un input, poder escribir en el y leer su valor", () => {
		// monta el componente
		render(<AddCategory onNewCategory={newCategory} />);
		// encuentra el input text
		const input: HTMLInputElement = screen.getByRole("textbox");
		// ingresa Saitama en el input
		fireEvent.input(input, { target: { value: "Saitama" } });
		// verifica si el valor de input es Saitama
		expect(input.value).toBe("Saitama");
	});

	test("should call the prop onNewCategory if the input has a value", () => {
		const inputValue = "Saitama";
		// creamos una funcion mock
		const onNewCategory = jest.fn();
		// montar el componente
		render(<AddCategory onNewCategory={onNewCategory} />);
		// para poder obtener el form, necesita agregarse un aria-label='form'
		const form = screen.getByRole("form");
		// obtenemos la referencia del input
		const input: HTMLInputElement = screen.getByRole("textbox");
		// ingresamos inputValue en el input a traves de un evento
		fireEvent.input(input, { target: { value: inputValue } });
		// ejecutar el submit del formulario
		fireEvent.submit(form);
		// despues de hacer submit, se vacia la caja de texto
		expect(input.value).toBe("");
    // evaluamos que onNewCategory fue llamada
    expect(onNewCategory).toHaveBeenCalled()
    // evaluamos que solo sea llamada 1 vez
    expect(onNewCategory).toHaveBeenCalledTimes(1)
    // evaluamos que haya sido llamada con el valor de la caja de texto
    expect(onNewCategory).toHaveBeenCalledWith(inputValue)
	});


  test("should not call onNewCategory if the input is empty", () => {
		const inputValue = "";
		// creamos una funcion mock
		const onNewCategory = jest.fn();
		// montar el componente
		render(<AddCategory onNewCategory={onNewCategory} />);
		// para poder obtener el form, necesita agregarse un aria-label='form'
		const form = screen.getByRole("form");
		// obtenemos la referencia del input
		const input: HTMLInputElement = screen.getByRole("textbox");
		// ingresamos inputValue en el input a traves de un evento
		fireEvent.input(input, { target: { value: inputValue } });
		// ejecutar el submit del formulario
		fireEvent.submit(form);
		// despues de hacer submit, se vacia la caja de texto
		expect(input.value).toBe("");
    // evaluamos que onNewCategory fue llamada
    expect(onNewCategory).not.toHaveBeenCalled()
    // evaluamos que solo sea llamada 1 vez
    expect(onNewCategory).not.toHaveBeenCalledTimes(1)
    // evaluamos que haya sido llamada con el valor de la caja de texto
    expect(onNewCategory).not.toHaveBeenCalledWith(inputValue)
	});
});
