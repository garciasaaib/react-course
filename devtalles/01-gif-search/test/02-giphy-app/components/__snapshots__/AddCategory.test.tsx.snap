// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddCategory /> Debe crear en snapshoot de AddCategory 1`] = `
<div>
  <form
    aria-label="form"
  >
    <input
      placeholder="Buscar gifs"
      type="text"
      value=""
    />
  </form>
</div>
`;

exports[`<AddCategory /> Debe crear un input en AddCategory 1`] = `
<div>
  <form
    aria-label="form"
  >
    <input
      placeholder="Buscar gifs"
      type="text"
      value=""
    />
  </form>
</div>
`;
